### PR TITLE
Conversions: Broaden Triggering

### PR DESCRIPTION
**Pull request by pjhampton**
## Description of new Instant Answer, or changes

This PR broadens the triggering and fixes a number of bugs. Due to the nature of these changes, we **might want to hit it with the internal query tool**.

### Significant changes
  - Handles natural lang queries such as:
    - "how many cm make a meter"
    - "how many hours are in a week"
    - "how many cms equal an inch"
    - "how many teaspoons equal a tablespoon"
    - "12888 minutes converted to hours"
  - handles awkward separators such as
    - "100 liters per gallon"
    - "celsius vs fahrenheit"
    - "36.0c is what in f"
    - "weeks = 3 months"
  - Moved lang triggers into its own file

### Minor changes:
  - Add calculator aliases, such as 'power|force|etc calculator'
  - Fixed triggering issue with grams
  - A load of quick win triggers
  - Removed smart matches so they don't end up in the server logs
  - Added abbreviated "tsp", "tsps" for teaspoon

### Future Discussion Points

  - Handling queries such as '3423 cm' (do we default?)
  - Handling queries such as 'inch converter' (do we default?)
  - Handling queries such as '6.3 km conversion' (do we default?)
  - Handling speed (Does math.js support?)
  - Cross base conversion - such as 13 foot squared to feet. [Google's Approach](https://www.google.co.uk/#q=13+foot+squared+to+feet)

## People to notify

@bsstoner

## Testing & Review
To be completed by Community Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [ ] Title follows correct format (Specifies Instant Answer + Purpose)
- [ ] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [ ] Instant Answer page is correctly filled out and contains:
    - The **Title** is appropriately named and formatted
    - The **IA topics** are present and appropriate
    - The **Description** is clear and coherent
    - **Source Name** exists if applicable
    - All **Example Queries** trigger on [Beta](https://beta.duckduckgo.com/)

**Code**
- [ ] Adheres to the [DuckDuckGo Style Guide](https://docs.duckduckhack.com/resources/code-style-guide.html)
- [ ] Behaviour is appropriately tested. If improvement, tests are adequately extended.
- [ ] There is no unnecessary files in place (such as editor config files)
- [ ] There is no API keys / secrets present
- [ ] Tests are passing (run `$ duckpan test <goodie_id>`)
    - Tester should report any failures

**Ready to merge?**

- [ ] Has this IA been deployed to and tested on [beta.duckduckgo.com](https://beta.duckduckgo.com/)?
- [ ] For larger commits, has this been approved be more than one community member?
- [ ] The number of reviews is appropriate for this type of PR
- [ ] The commit message is clear, coherent and fitting

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/conversions

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by moollaza** [comment]
@pjhampton I reverted the formatting change as it was going to introduce conflicts.

I think it would be ideal if we created a "Conversions" folder in the test dir and split the conversions tests across multiple files. That might also mitigate future merge conflicts.

File: lib/DDG/Goodie/Conversions.pm [hunk 1]
```diff
@@ -29,22 +29,27 @@ foreach my $type (@types) {
...
-                    'digital storage converter', 'duration conversion', 'duration converter', 'energy conversion', 'energy converter', 'force conversion', 'force converter',
-                    'power conversion', 'power converter', 'pressure conversion', 'pressure converter',
-                    'temperature conversion', 'temperature converter', 'volume conversion', 'volume converter');
+my @raw_lang_triggers = share('langTriggers.txt')->slurp;
+my @lang_triggers;
+
+# preprocess the lang based triggers
+foreach my $trig (@raw_lang_triggers) {
+  chomp($trig);
+  $trig =~ s/,//;
```
Comment (moollaza) [inline lib/DDG/Goodie/Conversions.pm]: can't we just remove these from the file?

File: lib/DDG/Goodie/Conversions.pm [hunk 2]
```diff
@@ -29,22 +29,27 @@ foreach my $type (@types) {
...
-                    'mass converter', 'angle conversion', 'angle converter', 'area conversion', 'area converter', 'digital storage conversion',
-                    'digital storage converter', 'duration conversion', 'duration converter', 'energy conversion', 'energy converter', 'force conversion', 'force converter',
-                    'power conversion', 'power converter', 'pressure conversion', 'pressure converter',
-                    'temperature conversion', 'temperature converter', 'volume conversion', 'volume converter');
+my @raw_lang_triggers = share('langTriggers.txt')->slurp;
+my @lang_triggers;
+
+# preprocess the lang based triggers
+foreach my $trig (@raw_lang_triggers) {
+  chomp($trig);
```
Comment (moollaza) [inline lib/DDG/Goodie/Conversions.pm]: we can chomp in the slurp: `slurp(chomp => 1)` 👍

**Comment by moollaza** [comment]
> Handling queries such as '3423 cm' (do we default?)
> Handling queries such as 'inch converter' (do we default?)
> Handling queries such as '6.3 km conversion' (do we default?)

Looks like Google triggers in all those cases. The last two show a pretty clear user intent so I feel those are obvious. I would imagine a search for a specific number + units is best served by a conversion as well. I think it would make sense to match Google's behaviour as users are likely expecting that.

@pjhampton this looks good to me

File: share/goodie/conversions/langTriggers.txt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+force conversion
+force converter
+force calculator
+power conversion
+power converter
+power calculator
+pressure conversion
+pressure converter
+pressure calculator
+temperature conversion
```
Comment (moollaza) [inline share/goodie/conversions/langTriggers.txt]: if we switch to yaml/json we could simplify this a bit by splitting the basic terms and generating the "conversion/converter/calculator" postfixes as needed:

```yaml
---
#generic
unit
metric
length
mass
...
---
# unique
online converter
metric conversion calculator
```

**Comment by moollaza** [comment]
@pjhampton Thanks! This LGTM :+1:

I've...
- simplified the trigger word loading
- added abbreviated triggers for teaspoon + test

**Comment by bsstoner** [comment]
This was released to production 🚀

**Pull request merged**